v2/lints/mozilla/lint_mp_pss_parameters_encoding_correct.go

```diff
@@ -0,0 +1,101 @@
+ */
+
+/************************************************
+
+https://www.mozilla.org/en-US/about/governance/policies/security-group/certs/policy/
+
+Section 5.1.1 RSA
+
+RSASSA-PSS with SHA-256, MGF-1 with SHA-256, and a salt length of 32 bytes.
+
+The encoded AlgorithmIdentifier MUST match the following hex-encoded bytes:
+
+304106092a864886f70d01010a3034a00f300d0609608648016503040201
+0500a11c301a06092a864886f70d010108300d0609608648016503040201
+0500a203020120
+
+RSASSA-PSS with SHA-384, MGF-1 with SHA-384, and a salt length of 48 bytes.
+
+The encoded AlgorithmIdentifier MUST match the following hex-encoded bytes:
+
+304106092a864886f70d01010a3034a00f300d0609608648016503040202
+0500a11c301a06092a864886f70d010108300d0609608648016503040202
+0500a203020130
+
+RSASSA-PSS with SHA-512, MGF-1 with SHA-512, and a salt length of 64 bytes.
+
+The encoded AlgorithmIdentifier MUST match the following hex-encoded bytes:
+
+304106092a864886f70d01010a3034a00f300d0609608648016503040203
+0500a11c301a06092a864886f70d010108300d0609608648016503040203
+0500a203020140
+************************************************/
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v2/lint"
+	"github.com/zmap/zlint/v2/util"
+)
+
+type rsaPssAidEncoding struct{}
+
+var RSASSAPSSAlgorithmIDToDER = [3][]byte{
+	// RSASSA-PSS with SHA-256, MGF-1 with SHA-256, salt length 32 bytes
+	{0x30, 0x41, 0x06, 0x09, 0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x0a, 0x30, 0x34, 0xa0, 0x0f, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01, 0x05, 0x00, 0xa1, 0x1c, 0x30, 0x1a, 0x06, 0x09, 0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x08, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01, 0x05, 0x00, 0xa2, 0x03, 0x02, 0x01, 0x20},
+	// RSASSA-PSS with SHA-384, MGF-1 with SHA-384, salt length 48 bytes
+	{0x30, 0x41, 0x06, 0x09, 0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x0a, 0x30, 0x34, 0xa0, 0x0f, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x02, 0x05, 0x00, 0xa1, 0x1c, 0x30, 0x1a, 0x06, 0x09, 0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x08, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x02, 0x05, 0x00, 0xa2, 0x03, 0x02, 0x01, 0x30},
+	// RSASSA-PSS with SHA-512, MGF-1 with SHA-512, salt length 64 bytes
+	{0x30, 0x41, 0x06, 0x09, 0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x0a, 0x30, 0x34, 0xa0, 0x0f, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x03, 0x05, 0x00, 0xa1, 0x1c, 0x30, 0x1a, 0x06, 0x09, 0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x08, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x03, 0x05, 0x00, 0xa2, 0x03, 0x02, 0x01, 0x40},
+}
+
+func (l *rsaPssAidEncoding) Initialize() error {
+	return nil
+}
+
+func (l *rsaPssAidEncoding) CheckApplies(c *x509.Certificate) bool {
+	return c.SignatureAlgorithmOID.Equal(util.OidRSASSAPSS)
```

Comment:
@cpu I may have missed it, but I didn't see anything in `rfc` or `cabf_br` that ensured that the the inner `signature` field matched the outer `signatureAlgorithm` field. Do you know of such a test?

The context is that the Mozilla Policy in https://www.mozilla.org/en-US/about/governance/policies/security-group/certs/policy/#51-Algorithms applies to both the `signatureAlgorithm` field of the Certificate, and the `signature` field of the TBSCertificate. This only tests the TBSCertificate (as does line 78)

I _suspect_ that this is perfectly fine for this PR, because it doesn't look like https://github.com/zmap/zcrypto/blob/master/x509/x509.go even exposes the ability to access the outer structure. That is, the `x509.Certificate` that is exported doesn't expose this information from https://github.com/zmap/zcrypto/blob/2177ed15981c76d13681ec89db9a1efd8e436230/x509/x509.go#L1458 , so this wouldn't be an easy fix. But just noting it in case we want to file a follow-up.

Reply:
> I didn't see anything in rfc or cabf_br that ensured that the the inner signature field matched the outer signatureAlgorithm field. Do you know of such a test?

@sleevi I don't know of any coverage for that within the existing lints :-( Thanks for raising this line of thought.

> This only tests the TBSCertificate (as does line 78)

Indeed, I think the older RSA parameter encoding lint is the same: https://github.com/zmap/zlint/blob/fd40f579253ea1ebfb18a585ab5cd8e7dcde61aa/v2/lints/rfc/lint_tbs_signature_rsa_encryption_parameter_not_null.go#L43-L44

> so this wouldn't be an easy fix. But just noting it in case we want to file a follow-up.

I think a follow-up would make the most sense. Probably expressed as two issues: one on the ZLint repo discussing the need to lint the `Certificate.signature`, and one on the ZCrypto repo discussing the need to export the outer structure. Would you like me to file those?

Reply:
> I think a follow-up would make the most sense. Probably expressed as two issues: one on the ZLint repo discussing the need to lint the `Certificate.signature`, and one on the ZCrypto repo discussing the need to export the outer structure. Would you like me to file those?

Done:
https://github.com/zmap/zcrypto/issues/213
https://github.com/zmap/zlint/issues/419

Reply:
Thank you!